### PR TITLE
adds regex check for RFC 1035

### PR DESCRIPTION
**Pull request by metalcyanide**

### CODE REVIEW

**Comment by metalcyanide** [comment]
build ako

File: internal/nodes/avi_model_l4_translator.go [hunk 1]
```diff
@@ -441,10 +441,17 @@ func getAutoFQDNForService(svcNamespace, svcName string) string {
 	if lib.GetL4FqdnFormat() == lib.AutoFQDNDefault {
 		// Generate the FQDN based on the logic: <svc_name>.<namespace>.<sub-domain>
 		fqdn = svcName + "." + svcNamespace + "." + subdomain
+
 	} else if lib.GetL4FqdnFormat() == lib.AutoFQDNFlat {
 		// Generate the FQDN based on the logic: <svc_name>-<namespace>.<sub-domain>
 		fqdn = svcName + "-" + svcNamespace + "." + subdomain
 	}
+	if !lib.CheckRFC1035(fqdn) {
```
Comment (chauhanshubham) [inline internal/nodes/avi_model_l4_translator.go:449]: The fqdn can be auto-generated from within AKO itself, which is not an explicit user input, in which case 
1. either this should not be enforced.
2. the auto-generated fqdn must be compliant at all times regardless of any user input.

Comment (metalcyanide) [reply internal/nodes/avi_model_l4_translator.go:449]: We need to follow the label standard RFC 1035. Either we can ask them to provide the svcName starting with an alphabetic character or we can add that tag "svc" when they don't. The other alternative is to always add the tag. Please suggest.

File: internal/nodes/avi_model_l4_translator.go [hunk 2]
```diff
@@ -441,10 +441,17 @@ func getAutoFQDNForService(svcNamespace, svcName string) string {
...
+		if lib.CheckRFC1035("svc" + fqdn) {
+			fqdn = "svc" + fqdn
```
Comment (akshayhavile) [inline internal/nodes/avi_model_l4_translator.go:451]: We can add warn message over here to notify user that fqdn has been changed. After line number 451.

File: internal/nodes/avi_model_l4_translator.go [hunk 3]
```diff
@@ -441,10 +441,17 @@ func getAutoFQDNForService(svcNamespace, svcName string) string {
...
+		}
+		utils.AviLog.Warnf("Provided fqdn: %s doesn't obey RFC 1035", fqdn)
```
Comment (akshayhavile) [inline internal/nodes/avi_model_l4_translator.go:453]: This message seems to be irrelevant as we are printing Warn message in  CheckRFC1035.

File: internal/lib/utils.go [hunk 4]
```diff
@@ -200,3 +201,26 @@ func Difference(a, b []string) []string {
...
+
+func CheckRFC1035(name string) bool {
+	RFCpattern := "^[a-z]([-a-z0-9]*[a-z0-9])?$"
+	maxlength := 63
+
+	if CheckConstraintsForRFC(name, RFCpattern, maxlength) {
+		return true
+	}
+
+	utils.AviLog.Warnf("Given string doesn't follow RFC 1035 constraints")
```
Comment (akshayhavile) [inline internal/lib/utils.go:224]: Warn message should be  `FQDN %s does not follow RFC 1035.`

**Comment by metalcyanide** [comment]
build ako

**Comment by metalcyanide** [comment]
build ako

File: internal/nodes/avi_model_l4_translator.go (hunk 2 above)
Comment (swathinsankaran) [inline internal/nodes/avi_model_l4_translator.go:451]: Shall we have a hyphen between string "svc" and fqdn so that it will become more readable?

File: internal/lib/utils.go [hunk 5]
```diff
@@ -200,3 +201,26 @@ func Difference(a, b []string) []string {
...
+
+	compliedRegex := regexp.MustCompile(pattern)
+	match := compliedRegex.Match([]byte(name))
+
+	return match
+}
+
+func CheckRFC1035(name string) bool {
+	RFCpattern := "^[a-z]([-a-z0-9]*[a-z0-9])?$"
+	maxlength := 63
```
Comment (swathinsankaran) [inline internal/lib/utils.go:218]: is the max length 255 or 63?

File: internal/lib/utils.go [hunk 6]
```diff
@@ -200,3 +201,26 @@ func Difference(a, b []string) []string {
 	}
 	return diff
 }
+
+func CheckConstraintsForRFC(name string, pattern string, maxlength int) bool {
+	if len(name) > maxlength {
+		utils.AviLog.Warnf("Given string is longer than expected. Maximum allowed length is: %d", maxlength)
```
Comment (swathinsankaran) [inline internal/lib/utils.go:207]: do we need to return a false here?

**Comment by metalcyanide** [comment]
build ako

**Comment by metalcyanide** [comment]
build ako

**Review by swathinsankaran — APPROVED** [review]

File: internal/lib/utils.go (hunk 6 above)
Comment (akshayhavile) [reply internal/lib/utils.go:207]: @metalcyanide : You can change log statement as "String %s has length more than maximum allowed length %d"

**Comment by metalcyanide** [comment]
build ako

**Comment by metalcyanide** [comment]
build ako

**Review by swathinsankaran — APPROVED** [review]

**Review by akshayhavile — APPROVED** [review]
